arm9/source/emulator.c

```diff
@@ -135,6 +136,42 @@ screen_talk(Device *d, Uint8 b0, Uint8 w)
 	return 1;
 }
 
+static Uint16
+get_entry(char *p, Uint16 len, const char *pathname, const char *basename, int fail_nonzero)
+{
+	struct stat st;
+	if(len < strlen(basename) + 7)
+		return 0;
+	if(stat(pathname, &st))
+		return fail_nonzero ? snprintf(p, len, "!!!! %s\n", basename) : 0;
+	else if(S_ISDIR(st.st_mode))
+		return snprintf(p, len, "---- %s\n", basename);
+	else if(st.st_size < 0x10000)
+		return snprintf(p, len, "%04x %s\n", (Uint16)st.st_size, basename);
+	else
+		return snprintf(p, len, "???? %s\n", basename);
+}
+
+static Uint16
+file_read_dir(char *dest, Uint16 len, DIR *dir, const char *filename)
+{
+	static char pathname[4096];
```

Comment:
4KB will unnecessarily eat a lot of stack on the NDS. I think 512 will be sufficient.